docs/source/en/agents.md

```diff
@@ -0,0 +1,570 @@
+{
+  "action": 'final_answer',
+  "action_input": "insert your final answer here"
+}
+
+Now begin!
+```
+
+The prompt has these parts:
+
+- Introduction: how the agent should behave, explanation of the concept of tools.
+- Description of all the tools. This is defined by a `<<tool_descriptions>>` token that is dynamically replaced at runtime with the tools defined/chosen by the user.
+    - The description is built based on the tool attributes `name`, `description`, `inputs` and `output_type`  with a simple `jinja2` template, that you can refine.
+- Clarifications on the output format
+- Introduction of the task at hand.
+
+This could be improved: for instance by adding explanations of the output format.
+
+For maximum flexibility, you can overwrite the whole prompt template as explained above by passing your custom prompt as an argument:
+
+```python
+from transformers import ReactAgent
+from transformers.tools import CalculatorTool
+
+agent = ReactAgent(llm_engine, tools = [CalculatorTool()], system_prompt="{your_custom_prompt}")
+```
+
+<Tip warning={true}>
+
+Please make sure to have the `<<all_tools>>` string defined somewhere in the `template` so that the agent can be aware 
+of the tools, it has available to it.
+
+</Tip>
+
+
+## Adding new tools
+
+In this section, we show how to create a new tool that can be added to the agent.
+
+### Creating a new tool and sharing it to the Hub
+
+We'll first start by creating a tool. We'll add the not-so-useful yet fun task of fetching the model on the Hugging Face
+Hub with the most downloads for a given task.
+
+We can do that with the following code:
+
+```python
+from huggingface_hub import list_models
+
+task = "text-classification"
+
+model = next(iter(list_models(filter=task, sort="downloads", direction=-1)))
+print(model.id)
+```
+
+For the task `text-classification`, this returns `'facebook/bart-large-mnli'`, for `translation` it returns `'google-t5/t5-base`.
+
+How do we convert this to a tool that the agent can leverage? All tools depend on the superclass `Tool` that holds the
+main attributes necessary. We'll create a class that inherits from it:
+
+```python
+from transformers import Tool
+
+class HFModelDownloadsTool(Tool):
+    pass
+```
+
+This class has a few needs:
+- An attribute `name`, which corresponds to the name of the tool itself. To be in tune with other tools which have a
+  performative name, we'll name it `model_download_counter`.
+- An attribute `description`, which will be used to populate the prompt of the agent.
+- `inputs` attribute. This is a dictionnary with keys `"type"`and `"description"`containing information to help the python interpreter make educated choices about the input,
+  and will allow for a gradio-demo to be spawned when we push our tool to the Hub.
+- An attribute `output_type` specifying the type of the ouput. 
+- A `__call__` method which contains the inference code. This is the code we've played with above!
+
+Here's what our class looks like now:
```

Comment:
```suggestion
The custom tool needs:

- An attribute `name`, which corresponds to the name of the tool itself. The name usually describes what the tool does. Since the code returns the model with the most downloads for a task, let's name is `model_download_counter`.
- An attribute `description` is used to populate the agent's system prompt.
- An `inputs` attribute, which is a dictionary with keys `"type"` and `"description"`. It contains information that helps the Python interpreter make educated choices about the input.
- An `output_type` attribute, which specifies the output type.
- A `__call__` method which contains the inference code to be executed.
```